src/io/import_off.cc

```diff
@@ -2,24 +2,234 @@
 #include "PolySet.h"
 #include "printutils.h"
 #include "AST.h"
-#ifdef ENABLE_CGAL
-#include "cgalutils.h"
-#endif
+#include <fstream>
+#include <boost/regex.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+// References:
+// http://www.geomview.org/docs/html/OFF.html
 
 std::unique_ptr<PolySet> import_off(const std::string& filename, const Location& loc)
 {
-#ifdef ENABLE_CGAL
-  CGAL_Polyhedron poly;
-  std::ifstream file(filename.c_str(), std::ios::in | std::ios::binary);
-  if (!file.good()) {
-    LOG(message_group::Warning, "Can't open import file '%1$s', import() at line %2$d", filename, loc.firstLine());
-  } else {
-    file >> poly;
-    file.close();
-    return CGALUtils::createPolySetFromPolyhedron(poly);
-  }
-#else
-  LOG(message_group::Warning, "OFF import requires CGAL, import() at line %2$d", filename, loc.firstLine());
-#endif // ifdef ENABLE_CGAL
-  return std::make_unique<PolySet>(3);
+  boost::regex ex_magic(R"(^(ST)?(C)?(N)?(4)?(n)?OFF( BINARY)? *)");
+  // XXX: are ST C N always in order?
+  boost::regex ex_cr(R"(\r$)");
+  boost::regex ex_comment(R"(\s*#.*$)");
+  boost::smatch results;
+
+  std::ifstream f(filename.c_str(), std::ios::in | std::ios::binary);
+
+  int lineno = 0;
+  std::string line;
+
+  auto AsciiError = [&](const auto& errstr){
+    LOG(message_group::Error, loc, "",
+    "OFF File line %1$s, %2$s line '%3$s' importing file '%4$s'",
+    lineno, errstr, line, filename);
+  };
+
+  auto getline_clean = [&](const auto& errstr){
+    lineno++;
+    std::getline(f, line);
+    if (f.eof()) {
+      AsciiError(errstr);
+      return false;
+    }
+    // strip DOS line endings
+    if (boost::regex_search(line, results, ex_cr))
+      line = line.erase(results.position(), results[0].length());
+    // strip comments
+    if (boost::regex_search(line, results, ex_comment))
+      line = line.erase(results.position(), results[0].length());
+    return true;
+  };
+
+  if (!f.good()) {
+    AsciiError("File error");
+    return std::make_unique<PolySet>(3);
+  }
+
+  bool got_magic = false;
+  // defaults
+  bool has_normals = false;
+  bool has_color = false;
+  bool has_textures = false;
+  bool has_ndim = false;
+  bool is_binary = false;
+  unsigned int dimension = 3;
+
+  while (line.length() == 0) {
```

Comment:
-> line.empty()
Could probably use empty() in several places below to make code a tiny bit clearer.

Reply:
OK, haven't used st:: stuff much in Haiku :-)

Reply:
> has no member named ‘empty’

Oh results[] is just the trait but not a real std::string, so I'd need `.str().empty()`… not shorter than `.length() > 0`.

Reply:
Right, I didn't spot that detail. Interesting tidbit about boost sub_match: It has a public `bool matched` member, so you can technically do `if (results[2].matched)`. Not really helping too much though.

Reply:
It's a little more obvious what is done yeah.
Now, I'm not really sure the various magic flags are always in that order though, if not I'd have to rewrite that part. For now I haven't found a file that didn't read though.